src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+            anyhow::anyhow!(
+                "HCSR04Sensor: failed to get `trigger_pin from configuration`: {:?}",
+                e
+            )
+        })?;
+
+        let echo_interrupt_pin = cfg
+            .get_attribute::<i32>("echo_interrupt_pin")
+            .map_err(|e| {
+                anyhow::anyhow!(
+                    "HCSR04Sensor: failed to get `echo_interrupt_pin` from configuration: {:?}",
+                    e
+                )
+            })?;
+
+        let timeout = cfg.get_attribute::<u32>("timeout_ms").map_or_else(
+            |e| match e {
+                AttributeError::KeyNotFound(_) => Ok(None),
+                _ => Err(anyhow::anyhow!(
+                    "HCSR04Sensor: error handling `timeout_ms` value: {:?}",
+                    e
+                )),
+            },
+            |v| Ok(Some(Duration::from_millis(v as u64))),
+        )?;
+
+        Ok(Arc::new(Mutex::new(HCSR04Sensor::new(
+            board,
+            trigger_pin,
+            echo_interrupt_pin,
+            timeout,
+        )?)))
+    }
+
+    fn new(
+        board: BoardType,
+        trigger_pin: i32,
+        echo_interrupt_pin: i32,
+        timeout: Option<Duration>,
+    ) -> anyhow::Result<HCSR04Sensor> {
+        // TODO: Move this into micro-rdk and unify with pin.rs
+        init_isr_alloc_flags(esp_idf_hal::interrupt::InterruptType::Iram.into());
+        enable_isr_service()?;
+
+        let notification = Notification::new();
+        let notifier = notification.notifier();
+
+        let mut sensor = Self {
+            _board: board,
+            trigger_pin: PinDriver::output(unsafe { AnyIOPin::new(trigger_pin) })?,
+            echo_interrupt_pin: PinDriver::input(unsafe { AnyIOPin::new(echo_interrupt_pin) })?,
+            timeout: timeout.unwrap_or(Duration::from_millis(50)),
+            interrupt_notification: notification,
+            isr_shared_state: Arc::new(IsrSharedState {
+                timestamp: 0.into(),
+                notifier,
+            }),
+        };
+
+        sensor.echo_interrupt_pin.set_pull(Pull::Down)?;
+
+        // Start the trigger pin high: the pulse is sent on the
+        // falling edge, so we can just go low immediately in
+        // `get_readings` to send it. As long as we don't get another
+        // `get_readings` request within 10us of the prior one
+        // completing, the pin will be high long enough to trigger the
+        // pulse.
+        sensor.trigger_pin.set_high()?;
+
+        sensor
+            .echo_interrupt_pin
+            .set_interrupt_type(InterruptType::AnyEdge)?;
+
+        unsafe {
+            esp!(gpio_isr_handler_add(
+                echo_interrupt_pin,
+                Some(Self::subscription_interrupt),
+                Arc::as_ptr(&sensor.isr_shared_state) as *mut _,
+            ))?;
+        }
+
+        Ok(sensor)
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.intr_srv"]
```

Comment:
I copied this verbatim from how `pin.rs` is set up, but I'm not 100% sure whether it is required or not.

Reply:
this is just the section where to code is going , the important part is that it is going in  the iram (maybe not needed for low-frequency interrupt)

Reply:
Yes, I'm clear on the link_section part; it was more the second part (whether it should be or is required to be in iram or not) that I was (and remain) unsure about.

Reply:
i don't think  it is since it is not time critical or will be often called